irc-proto/src/prefix.rs

```diff
@@ -0,0 +1,204 @@
+//! A module providing an enum for a message prefix.
+use std::borrow::ToOwned;
+use std::string;
+use std::str::FromStr;
+use std::fmt;
+
+/// The Prefix indicates "the true origin of the message", according to the server.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Prefix {
+    /// servername, e.g. collins.mozilla.org
+    ServerName(String),
+    /// nickname [ ["!" username] "@" hostname ]
+    /// i.e. Nickname(nickname, username, hostname)
+    /// Any of the strings may be ""
+    Nickname(String, String, String),
+}
+
+impl Prefix {
+    /// Creates a prefix by parsing a string.
+    ///
+    /// # Example
+    /// ```
+    /// # extern crate irc;
+    /// # use irc::client::prelude::*;
+    /// # fn main() {
+    /// Prefix::new_from_str("nickname!username@hostname");
+    /// Prefix::new_from_str("example.com");
+    /// # }
+    /// ```
+    pub fn new_from_str(s: &str) -> Prefix {
+        #[derive(Copy, Clone, Eq, PartialEq)]
+        enum Active {
+            Name = 0,
+            User = 1,
+            Host = 2,
+        }
+
+        let mut name = String::new();
+        let mut user = String::new();
+        let mut host = String::new();
+        let mut active = Active::Name;
+
+        for c in s.chars() {
+            match c {
+                // We consider the '.' to be a ServerName except if a ! has already
+                // been encountered.
+                '.' if active == Active::Name => return Prefix::ServerName(s.to_owned()),
+
+                '!' if active == Active::Name => {
+                    active = Active::User;
+                },
+
+                // The '@' is not special until we've started the username
+                // portion
+                '@' if active == Active::User => {
+                    active = Active::Host;
+                },
+
+                _ => {
+                    // Push onto the latest buffer
+                    match active {
+                        Active::Name => &mut name,
+                        Active::User => &mut user,
+                        Active::Host => &mut host,
+                    }.push(c)
+                }
+            }
+        }
+
+        Prefix::Nickname(name, user, host)
+    }
+}
+
+/// This implementation never returns an error and is isomorphic with `Display`.
+impl FromStr for Prefix {
+    type Err = string::ParseError;
```

Comment:
I think the unit type `()` usually is used when this can't return an error. (Eventually it instead will be the never type `!`.)

Reply:
This was recommended to me as it's basically `!`, and `let s: Prefix = s.parse().unwrap()` will continue to be safe unless the `type Err` is changed.

I'd like to keep the current code.

Reply:
`string::ParseError` is definitely not `!` (which should be read as "diverges"/"never returns"/etc.). As it stands, the type says that it might return a `ParseError`, even if you never return it. Ideally, it should be a `!`, but since that is not currently allowed in Rust, the current convention for non-erroring things is to use `()`.

Reply:
`string::ParseError` is an empty enum, so there's no way to a value of type `ParseError` in safe rust, which means the function can only possibly return `Ok(s)`.

Changed to `type Err = ()`.

Reply:
Ah, sorry, I missed that part (which is itself an argument to just write `()`, to be honest).

Reply:
Now that I've read the docs for `std::string::ParseError` — I didn't realize we had uninhabited error types in `std` — I think it does sound more appropriate than `()`.